### PR TITLE
selfhost/typechecker: Fix function return typecheck

### PR DESCRIPTION
**Pull request by adleroliveira**
This fix some bugs on function return type typeckecking and adds a PASS for the following tests:

**tests/typechecker/return_value_type_mismatch.jakt
tests/typechecker/function_return_type_with_fat_arrow.jakt**

Previously, when going through the typechecking namespaces declaration, the current_function_id was not being updated so the typechecker didn't got the return type hint correctly.

Also, in order for this to work correctly I created the function unify_with_type based on the same function on the Rust compiler. This function properly check for type compatibility.

There is a unify function that just compares type_id's and is being used on several places as a placeholder. We should probably start replacing it for the new one.

Before:
273 passed
60 failed
7 skipped

After
275 passed
58 failed
7 skipped

### CODE REVIEW

File: selfhost/typechecker.jakt [hunk 1]
```diff
@@ -1431,6 +1431,22 @@ struct Typechecker {
...
+                let generic_inferences: [String:String] = [:]
+                if .check_types_for_compat(lhs_type_id: expected_type!, rhs_type_id: found_type, generic_inferences, span) {
+                    return found_type
+                }
+
+                return .substitute_typevars_in_type(type_id: found_type, generic_inferences)
+        } else {
+            return found_type
+        }
+    }
```
Comment (awesomekling) [inline selfhost/typechecker.jakt:1448]: Please use 4 spaces for indentation.

File: selfhost/typechecker.jakt [hunk 2]
```diff
@@ -2697,7 +2715,7 @@ struct Typechecker {
         }
 
         parsed_function.must_instantiate = true
-
+
```
Comment (awesomekling) [inline selfhost/typechecker.jakt:2718]: Unrelated whitespace change.

File: selfhost/typechecker.jakt [hunk 3]
```diff
@@ -2896,7 +2914,7 @@ struct Typechecker {
                                 "Type mismatch: expected ‘{}’, but got ‘{}’"
                                 .type_name(TypeId::from_string(seen_type_id_string.value()))
                                 .type_name(rhs_type_id)
-                            )
+                            ),
```
Comment (awesomekling) [inline selfhost/typechecker.jakt:2917]: Unrelated change. Also comma is not required here.
There are a whole bunch of unrelated changes like this in the commit, please remove them :)

File: selfhost/typechecker.jakt [hunk 4]
```diff
@@ -3247,7 +3265,7 @@ struct Typechecker {
                     .check_types_for_compat(
                         lhs_type_id: checked_block.yielded_type.value()
                         rhs_type_id: type_
-                        generic_enferences: generic_inferences
+                        generic_inferences: generic_inferences
```
Comment (awesomekling) [inline selfhost/typechecker.jakt:3268]: Unrelated change. (Also, the fact that this worked before is probably a bug on its own!)

**Review by awesomekling — CHANGES_REQUESTED** [review]

File: selfhost/typechecker.jakt [hunk 5]
```diff
@@ -4175,7 +4193,10 @@ struct Typechecker {
         }
         SingleQuotedString(val, span) => CheckedExpression::CharacterConstant(val, span)
         SingleQuotedByteString(val, span) => CheckedExpression::ByteConstant(val, span)
-        QuotedString(val, span) => CheckedExpression::QuotedString(val, span)
+        QuotedString(val, span) => {
+            .unify_with_type(found_type: TypeId::from_string("0_13"), expected_type: type_hint, rhs_span: span)
```
Comment (awesomekling) [inline selfhost/typechecker.jakt:4197]: What is the significance of `"0_13"` here? There is no comment, or explanation in the commit message.

Comment (adleroliveira) [reply selfhost/typechecker.jakt]: replaced it for builtin(BuiltinType::String) which represents the same thing. Much more readable now.

Comment (awesomekling) [reply selfhost/typechecker.jakt]: Great!
